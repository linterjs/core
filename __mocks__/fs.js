const realFs = jest.requireActual("fs");
const { vol } = jest.requireActual("memfs");
const { ufs } = jest.requireActual("unionfs");
ufs.use(realFs).use(vol);
module.exports = ufs;
