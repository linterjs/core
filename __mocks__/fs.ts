const fs = require.requireActual("fs");

let files = Object.create(null);

function __clearFiles() {
  files = Object.create(null);
}

function __setFiles(newFiles: object) {
  Object.entries(newFiles).forEach(([path, content]) => {
    files[path] = content;
  });
}

const mockFs = {
  ...fs,
  __clearFiles,
  __setFiles,
  readFileSync: jest.fn((path, options) => {
    if (path in files) {
      return files[path];
    }

    fs.readFileSync(path, options);
  }),
  writeFileSync: jest.fn((path, data, options) => {
    if (path in files) {
      files[path] = data;
    } else {
      fs.writeFileSync(path, data, options);
    }
  })
};

module.exports = mockFs;
