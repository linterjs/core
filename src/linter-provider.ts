import readPkgUp = require("read-pkg-up");
import { oneLine } from "common-tags";
import {
  LinterProviderLoadError,
  ModuleNotLinterProviderError,
  NoLinterProvidersError,
} from "./errors";
import { LinterFactory } from "./linter-adapter";
import { logger } from "./logger";
import requireRelative = require("require-relative");

export interface LinterProvider {
  factory: LinterFactory;
  name: string;
  supportedExtensions: string[];
}

interface LinterProviderModule {
  default: LinterProvider;
}

function isLinterProviderModule(
  linterProviderModule: any,
): linterProviderModule is LinterProviderModule {
  return (
    linterProviderModule.default &&
    typeof linterProviderModule.default.factory === "function"
  );
}

export function loadLinterProvidersFromFile(): Set<LinterProvider> {
  const { packageJson } = readPkgUp.sync({ normalize: false }) ?? {};

  const dependencies = packageJson?.dependencies ?? {};
  const devDependencies = packageJson?.devDependencies ?? {};
  const optionalDependencies = packageJson?.optionalDependencies ?? {};

  const linterProviderModuleNames = Object.keys({
    ...dependencies,
    ...devDependencies,
    ...optionalDependencies,
  }).filter(
    (dependency) =>
      dependency.startsWith("@linter/provider-") ||
      dependency.startsWith("linter-provider-") ||
      (dependency.startsWith("@") && dependency.includes("/linter-provider-")),
  );

  const loadedlinterProviders = linterProviderModuleNames.reduce(
    (loadedLinterProviderAccumulator, linterProviderModuleName) => {
      let linterProviderModule: LinterProviderModule | undefined;
      try {
        const modulePath = require.resolve(linterProviderModuleName, {
          paths: [process.cwd()],
        });
        linterProviderModule = require(modulePath);
      } catch (error) {
        logger.debug(oneLine`
          Could not import "${linterProviderModuleName}",
          is it properly installed?
        `);
        logger.trace(error);

        if (!optionalDependencies[linterProviderModuleName]) {
          logger.debug(
            "Linter provider module is not optional, throwing error",
          );
          throw new LinterProviderLoadError(linterProviderModuleName);
        }
      }

      if (
        linterProviderModule &&
        isLinterProviderModule(linterProviderModule)
      ) {
        loadedLinterProviderAccumulator.add(linterProviderModule.default);
      } else if (linterProviderModule) {
        throw new ModuleNotLinterProviderError(linterProviderModuleName);
      }

      return loadedLinterProviderAccumulator;
    },
    new Set<LinterProvider>(),
  );

  if (loadedlinterProviders.size === 0) {
    const error = new NoLinterProvidersError();
    logger.debug(error.message);
    throw error;
  }

  return loadedlinterProviders;
}

// XXX: Add more functionality, enable/disable linters?
