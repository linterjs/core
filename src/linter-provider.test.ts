import fs from "jest-plugin-fs";
import {
  LinterAdapter,
  LinterAdapterFormat,
  LinterAdapterLint
} from "./linter-adapter";
import { LinterProvider, loadLinterProvidersFromFile } from "./linter-provider";

jest.mock("@linter/provider-eslint");
jest.mock("@zimme/linter-provider-eslint");
jest.mock("fs", () => require("jest-plugin-fs/mock"));
jest.mock("linter-provider-eslint");

beforeAll(() => {
  fs.root = process.cwd();
});

afterEach(() => {
  fs.restore();
  jest.resetModules();
});

describe("Load linter providers", () => {
  test("when a package.json can't be found", () => {
    fs.mock();

    expect(() => {
      const linterProviders = loadLinterProvidersFromFile();
    }).toThrowErrorMatchingSnapshot();
  });

  test("when an installed linter provider module can't be imported", () => {
    fs.mock({
      "package.json": JSON.stringify({
        devDependencies: { "@linter/provider-nonexistent": "1.0.0" }
      })
    });

    expect(() => {
      const linterProviders = loadLinterProvidersFromFile();
    }).toThrowErrorMatchingSnapshot();
  });

  test("with a linter provider installed", () => {
    fs.mock({
      "package.json": JSON.stringify({
        dependencies: {
          "@linter/provider-eslint": "1.0.0"
        }
      })
    });

    const eslintLinterProvider = require("@linter/provider-eslint").default;
    const loadedLinterProviders = loadLinterProvidersFromFile();
    expect(loadedLinterProviders).toEqual(new Set([eslintLinterProvider]));
  });

  test("with an installed linter provider and a missing optional linter provider", () => {
    fs.mock({
      "package.json": JSON.stringify({
        devDependencies: { "@linter/provider-eslint": "1.0.0" },
        optionalDependencies: { "@linter/provider-nonexistent": "1.0.0" }
      })
    });

    const eslintLinterProvider = require("@linter/provider-eslint").default;
    const loadedLinterProviders = loadLinterProvidersFromFile();
    expect(loadedLinterProviders).toEqual(new Set([eslintLinterProvider]));
  });

  test("using all linter provider package naming convensions", () => {
    fs.mock({
      "package.json": JSON.stringify({
        devDependencies: {
          "@linter/provider-eslint": "1.0.0",
          "@zimme/linter-provider-eslint": "1.0.0",
          "linter-provider-eslint": "1.0.0"
        }
      })
    });

    const eslintLinterProvider = require("@linter/provider-eslint").default;
    const loadedLinterProviders = loadLinterProvidersFromFile();
    expect(loadedLinterProviders).toEqual(new Set([eslintLinterProvider]));
  });

  test("when the linter provider package doesn't export a default linter provider", () => {
    fs.mock({
      "package.json": JSON.stringify({
        devDependencies: {
          "linter-provider-not-a-provider": "1.0.0"
        }
      })
    });

    expect(() => {
      const loadedLinterProviders = loadLinterProvidersFromFile();
    }).toThrowErrorMatchingSnapshot();
  });
});
