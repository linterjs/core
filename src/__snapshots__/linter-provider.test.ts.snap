// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load linter providers when a package.json can't be found 1`] = `"No linter providers found. Please install at least one linter provider package."`;

exports[`Load linter providers when an installed linter provider module can't be imported 1`] = `"There was a problem importing the linter provider module: \\"@linter/provider-nonexistent\\"."`;

exports[`Load linter providers when the linter provider package doesn't export a default linter provider 1`] = `"The imported module is not a linter provider: \\"linter-provider-not-a-provider\\""`;
